src/hotspot/cpu/aarch64/templateTable_aarch64.cpp

```diff
@@ -3069,14 +3076,15 @@ void TemplateTable::fast_storefield(TosState state)
 
   // access constant pool cache
   __ load_field_entry(r2, r1);
+
+  // Must prevent reordering of the following cp cache loads with bytecode load
+  __ membar(MacroAssembler::LoadLoad);
+
   __ push(r0);
   // R1: field offset, R2: TOS, R3: flags
   load_resolved_field_entry(r2, r2, r0, r1, r3);
```

Comment:
It is useless to use r0 here, so can we change it to noreg and eliminate the use of push(r0)/pop(r0)?

Reply:
I also noticed this today and I am testing the following change:
[unnecessary-tos-load-v2.diff.txt](https://github.com/openjdk/jdk/files/14758318/unnecessary-tos-load-v2.diff.txt)
@matias9927 : Maybe you can this change while you are on it? I think we should fix this as the 8-byte push/pop would violate the aarch64 and riscv64 ABI where a stack alignment of 16bytes is required.

Reply:
Noted! I haven't been able to replicate the crash yet, but I think it will be worth moving forward with this patch as a matter of correctness.

Reply:
Note that `noreg` is currently not properly handled / considered in `load_resolved_field_entry`. And it doesn't look nice to me to only check if `tos_state` with `noreg` in this function. That's why I replaced call to `load_resolved_field_entry` with seperate loads.

Reply:
I didn't notice your message until just now, but you're right. I would prefer to use `load_resolved_field_entry` when possible but there are places where fields are loaded individually so I think it's fine.

Reply:
All right. Another place in `TemplateTable::fast_accessfield` which might be worth turning into a `load_resolved_field_entry` call with `noreg` for `tos_state` for consistency:

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/templateTable_aarch64.cpp#L3169-L3170
[2] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/templateTable_riscv.cpp#L3139-L3140

Reply:
I think this change should be limited to fixing the membar issue and the push/pop issue and limit refactoring to a new change if more is wanted, but maybe this could call load_resolved_field_entry with noreg for tos-state too.